Refactor RgbdRenderer

thduynguyen opened the pull request:
Refactor RgbdRenderer to abstract class with a specific RgbdRendererVTK implementation. This is to allow a new implementation of RgbdRenderer, using for example Godot.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7611)
<!-- Reviewable:end -->

thduynguyen commented:
+@kunimatsu-tri for feature review please!

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0Lb0I0uCLbjnQ3bG0g:bcvlne7)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Completed the first pass. PTAL.

---

Reviewed 9 of 9 files at r1.
Review status: 2 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.h, line 19 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Lbw3wpgNf-Gmvbmzj:-L0Lbw3xVxXTkgKl_ORp:bh8j19f) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer.h#L19)):*
> ```objc
> namespace sensors {
> 
> /// Abstract interface of RGB-D renderers that renders RGB, depth and label
> ```

BTW, nit `that renders` -> `that render`.

---

*[systems/sensors/rgbd_renderer.h, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LcSPdxvcj0vI6ebih:-L0LcSPdxvcj0vI6ebii:b7637f0) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer.h#L176)):*
> ```objc
>   virtual void DoRenderLabelImage(ImageLabel16I* label_image_out) const = 0;
> 
>  protected:
> ```

`protected:` section should be placed before `private:` section.
https://google.github.io/styleguide/cppguide.html#Declaration_Order

---

*[systems/sensors/rgbd_renderer_vtk.cc, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LxD9g4R4RZSfG4Za-:-L0LxD9g4R4RZSfG4Za0:bf8drk2) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer_vtk.cc#L131)):*
> ```cpp
>   float CheckRangeAndConvertToMeters(float z_buffer_value) const;
> 
>   Parent& parent_;
> ```

Hmm, I'm not sure if this is a good design or now.

An alternative design would be that `RgbdRendererVTK` owns those variables instead of `RgbdRenderer`. In that case, creating a new class something like `RenderingConfig` that has accessors for `width`, `height`, `fov_y`, `z_near` and `z_far` might mitigate your pain to write many wrapper functions.

---

*[systems/sensors/rgbd_renderer_vtk.h, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LfGNigcb32sEgMIBQ:-L0LfGNigcb32sEgMIBR:bcpvxbb) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer_vtk.h#L11)):*
> ```objc
> /// An RgbdRenderer implementation using VTK
> class RgbdRendererVTK final : public RgbdRenderer {
>   using Base = RgbdRenderer;
> ```

Would it be possible to put this in `private` section for the clarity?

---

*[systems/sensors/rgbd_renderer_vtk.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LdpmSIOUsPFIWPsUp:-L0LdpmSIOUsPFIWPsUq:b-6urt7r) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer_vtk.h#L38)):*
> ```objc
>   void DoRenderLabelImage(ImageLabel16I* label_image_out) const override;
> 
>  private:
> ```

BTW, `private:` is declared twice.

---

*[systems/sensors/test/rgbd_renderer_test.h, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Lrx2Pf5vCyzX4Xhhe:-L0Lrx2Pf5vCyzX4Xhhf:bcf9gwd) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L1)):*
> ```objc
> #pragma once
> ```

(for the file name)

Probably `rgbd_renderer_test_util.h` is better since this is not a test but an util for tests.

---

*[systems/sensors/test/rgbd_renderer_test.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LoorutEWpi97P4ReB:-L0LoorutEWpi97P4ReC:ba9tqoc) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L13)):*
> ```objc
> namespace systems {
> namespace sensors {
> namespace {
> ```

How about replacing unnamed `namespace` with e.g. `namespace test` , since this part is moved from `.cc`? 

---

*[systems/sensors/test/rgbd_renderer_test.h, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Ltyx7difVpjLBzGNh:-L0Ltyx7difVpjLBzGNi:b-by768h) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L15)):*
> ```objc
> namespace {
> 
> using Eigen::Isometry3d;
> ```

Probably, it's better to avoid using `using` for this purpose in header files.

---

*[systems/sensors/test/rgbd_renderer_test.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LplSQjsB3j4yhy-i_:-L0LplSQjsB3j4yhy-ia:br2hw2) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L17)):*
> ```objc
> using Eigen::Isometry3d;
> 
> const int kWidth = 640;
> ```

We should avoid defining static variables in header files: 
https://google.github.io/styleguide/cppguide.html#Unnamed_Namespaces_and_Static_Variables

ditto for the variables below.

---

*[systems/sensors/test/rgbd_renderer_test.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LnYftTG1UAholus_x:-L0LnYftTG1UAholus_y:b-vp7z7n) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L46)):*
> ```objc
> 
> // This suite tests RgbdRenderer.
> template <class RGBD_RENDERER>
> ```

"Template parameters should follow the naming style for their category."
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

In this case, it would be `RgbdRenderer`. However, this is confusing since the class `RgbdRenderer` already exists.

So, how about simply using `T` instead?

---

*[systems/sensors/test/rgbd_renderer_test.h, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LuR4kR5hcjUgB9Bp7:-L0LuR4kR5hcjUgB9Bp8:b-pqc18v) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L102)):*
> ```objc
>   }
> 
>   void SetUp() override {}
> ```

Sorry, this is my mistake. Could you get rid of this?

---

*[systems/sensors/test/rgbd_renderer_test.h, line 105 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LuZAmFAl-AbrRSOeX:-L0LuZAmFAl-AbrRSOeY:b-q4mjve) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L105)):*
> ```objc
> 
>   // All tests on this class must invoke this first.
>   void SetUp(const Eigen::Isometry3d& X_WR, bool add_terrain = false) {
> ```

And, could you replace `SetUp` with e.g. `Init`?

Because I got a review comment saying " this is confusing since you're not using `SetUp` function from `testing::Test`."

Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0LyXAVTHS9P4gkcMPq:b-gv3nct)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Reviewed 1 of 9 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[systems/sensors/BUILD.bazel, line 491 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Lzx8kHQ9Fs4tRccK3:-L0Lzx8kHQ9Fs4tRccK4:b-x3gtbh) ([raw file](https://github.com/robotlocomotion/drake/blob/9aabcf7da511902760120fb23e670ab3b1fb0325/systems/sensors/BUILD.bazel#L491)):*
> ```Python
>     # Mitigates driver-related issues when running under `bazel test`. For more
>     # information, see #7004.
>     local = 1,
> ```

Thanks for catching this!

---

*[systems/sensors/test/rgbd_renderer_test.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LoorutEWpi97P4ReB:-L0M-YLvpOTYxagkhRWV:b-aziv63) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test.h#L13)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

How about replacing unnamed `namespace` with e.g. `namespace test` , since this part is moved from `.cc`? 
</blockquote></details>

Never mind. You've done that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
All comments addressed. PTAL.

---

Reviewed 1 of 10 files at r1, 5 of 9 files at r3.
Review status: 4 of 8 files reviewed at latest revision, 13 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.h, line 19 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Lbw3wpgNf-Gmvbmzj:-L0Qq9YSN5aG5bhWq2Qx:b-886gsx) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer.h#L19)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

BTW, nit `that renders` -> `that render`.
</blockquote></details>

Fixed

---

*[systems/sensors/rgbd_renderer.h, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LcSPdxvcj0vI6ebih:-L0QqFvrlMxWgefXRMoP:b-886gsx) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer.h#L176)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

`protected:` section should be placed before `private:` section.
https://google.github.io/styleguide/cppguide.html#Declaration_Order
</blockquote></details>

Fixed

---

*[systems/sensors/rgbd_renderer_vtk.cc, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LxD9g4R4RZSfG4Za-:-L0PldIqsn05erJHDB-Q:bftyv5r) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer_vtk.cc#L131)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Hmm, I'm not sure if this is a good design or now.

An alternative design would be that `RgbdRendererVTK` owns those variables instead of `RgbdRenderer`. In that case, creating a new class something like `RenderingConfig` that has accessors for `width`, `height`, `fov_y`, `z_near` and `z_far` might mitigate your pain to write many wrapper functions.
</blockquote></details>

I like the idea of `RenderingConfig`, but don't see why this is not a good design. If `RgbdRendererVTK` owns these variables, we will need to virtualize all the accessor functions, which is not really necessary for small variables, e.g. width/height, that all derived renderers need. We will still need this parent_ variable in Impl to access those properties in `RgbdRendererVTK` anyway. One option is to return the full `RenderingConfig` instead of its members. However, the `ColorPalette` member is large, so we should either (1) return a const ref to the `RenderingConfig`, or (2) return the `RenderingConfig` by value but make the color palette a static const or const ref singleton. (1) makes more sense to me; however, it requires all implementations of `RgbdRenderer` to have that member somewhere, so putting that in the base class still sounds more practical, although I can see the OOP desire to keep the abstract interface free of members.

Idea?

---

*[systems/sensors/rgbd_renderer_vtk.cc, line 132 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Q5rDAGGML6cTsmB7c:-L0Q5rDAGGML6cTsmB7d:bdrnlq0) ([raw file](https://github.com/robotlocomotion/drake/blob/9aabcf7da511902760120fb23e670ab3b1fb0325/systems/sensors/rgbd_renderer_vtk.cc#L132)):*
> ```cpp
> 
>   Parent* parent_ = nullptr;
> ```

Have discussed offline with @kunimatsu. The problem is deeper than this. We need a major redesign to split this into different RGB and Depth and Label renderers for different corresponding types of cameras, which need different sets of properties. One `RenderingConfig` class is not enough to handle all these cases. For example, a depth renderer needs not only `z_near`, `z_far` but also clipping planes to handle too close/too far objects, a RGB renderer requires lighting/environment settings which other renderers don't need, and the color palette is only needed for the Label renderer.

For simplicity and minimal code changes, I'll put `RenderingConfig` and also `ColorPalette` in the base class for now. With that decision, the `Parent` pattern makes sense as `Impl` implements its parent so it needs access to all members of the parent anyway.

---

*[systems/sensors/rgbd_renderer_vtk.h, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LfGNigcb32sEgMIBQ:-L0QqLxwMK7K22Q2YDfw:ba3dfob) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer_vtk.h#L11)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Would it be possible to put this in `private` section for the clarity?
</blockquote></details>

Removed

---

*[systems/sensors/rgbd_renderer_vtk.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LdpmSIOUsPFIWPsUp:-L0QqOyLtwv1FU7a_eb0:b-886gsx) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer_vtk.h#L38)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

BTW, `private:` is declared twice.
</blockquote></details>

Fixed

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Lrx2Pf5vCyzX4Xhhe:-L0QulxTur-qnyCmUGk6:b-886gsx) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L1)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

(for the file name)

Probably `rgbd_renderer_test_util.h` is better since this is not a test but an util for tests.
</blockquote></details>

Fixed

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LoorutEWpi97P4ReB:-L0QqVDot_fbmhLtJQuG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L13)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Never mind. You've done that.
</blockquote></details>

Done.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Ltyx7difVpjLBzGNh:-L0QuhCjnmc1oY7NZYWj:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L15)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Probably, it's better to avoid using `using` for this purpose in header files.
</blockquote></details>

Fixed.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LplSQjsB3j4yhy-i_:-L0Qq_d0WaW8J77W8zkK:b-zctniy) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L17)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

We should avoid defining static variables in header files: 
https://google.github.io/styleguide/cppguide.html#Unnamed_Namespaces_and_Static_Variables

ditto for the variables below.
</blockquote></details>

Move them inside the test class. This makes more sense as the expected constant values inside the test class depend on these as inputs.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LnYftTG1UAholus_x:-L0QqQPhWqK4QQKED0oj:b-lu9ppk) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L46)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

"Template parameters should follow the naming style for their category."
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

In this case, it would be `RgbdRenderer`. However, this is confusing since the class `RgbdRenderer` already exists.

So, how about simply using `T` instead?


</blockquote></details>

`T` is not informative. Change to `Renderer`.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LuR4kR5hcjUgB9Bp7:-L0QqrMXfuSC-x7tmkvc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L102)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Sorry, this is my mistake. Could you get rid of this?
</blockquote></details>

Done.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 105 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LuZAmFAl-AbrRSOeX:-L0QqsbmcbMI0FMw5Vhx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L105)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

And, could you replace `SetUp` with e.g. `Init`?

Because I got a review comment saying " this is confusing since you're not using `SetUp` function from `testing::Test`."

Thanks.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0Qus1dFXCdAchT3OKE:bymsvaa)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[systems/sensors/test/rgbd_renderer_test_util.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LnYftTG1UAholus_x:-L0QxioAcOIqAFsxhK1a:b-nqhkfu) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/test/rgbd_renderer_test_util.h#L46)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

`T` is not informative. Change to `Renderer`.
</blockquote></details>

OK, that's better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few minor BTW comments. Thank you for making this change!

---

Reviewed 1 of 10 files at r1, 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.h, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0QzYODXV67b7gdbt9M:-L0QzYODXV67b7gdbt9N:bbqmohv) ([raw file](https://github.com/robotlocomotion/drake/blob/376a90dc4748fb3122290f0edac06e6576e2ab45/systems/sensors/rgbd_renderer.h#L145)):*
> ```objc
>   void RenderLabelImage(ImageLabel16I* label_image_out) const;
> 
>   /// Returns the configuration object of this renderer
> ```

BTW, missing a period at the end of the sentence. It would be nice to be consistent to other comments.

ditto for other comments that don't have a period.

---

*[systems/sensors/rgbd_renderer.h, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Qyg-76MSF_VnEXfMq:-L0Qyg-76MSF_VnEXfMr:bxe2pbt) ([raw file](https://github.com/robotlocomotion/drake/blob/376a90dc4748fb3122290f0edac06e6576e2ab45/systems/sensors/rgbd_renderer.h#L162)):*
> ```objc
> 
>   /// The color palette for sky, terrain colors and ground truth label rendering
>   /// TODO: This is a world's property (colors for each object/segment)
> ```

BTW, TODO comment should be followed by responsible person's name(s), e.g. `TODO(thduynguyen, kunimatsu-tri) This is a ...`

---

*[systems/sensors/rgbd_renderer_vtk.cc, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0LxD9g4R4RZSfG4Za-:-L0R04CeOU5GiDX_ryaL:b-v6rj4u) ([raw file](https://github.com/robotlocomotion/drake/blob/38031826d2789cfd791f1b7ea0931c68ae61f5af/systems/sensors/rgbd_renderer_vtk.cc#L131)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

I like the idea of `RenderingConfig`, but don't see why this is not a good design. If `RgbdRendererVTK` owns these variables, we will need to virtualize all the accessor functions, which is not really necessary for small variables, e.g. width/height, that all derived renderers need. We will still need this parent_ variable in Impl to access those properties in `RgbdRendererVTK` anyway. One option is to return the full `RenderingConfig` instead of its members. However, the `ColorPalette` member is large, so we should either (1) return a const ref to the `RenderingConfig`, or (2) return the `RenderingConfig` by value but make the color palette a static const or const ref singleton. (1) makes more sense to me; however, it requires all implementations of `RgbdRenderer` to have that member somewhere, so putting that in the base class still sounds more practical, although I can see the OOP desire to keep the abstract interface free of members.

Idea?
</blockquote></details>

BTW, discussed with f2f and resolved.

---

*[systems/sensors/rgbd_renderer_vtk.cc, line 132 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Q5rDAGGML6cTsmB7c:-L0R-rvBOZKPWyCac3vh:bpqp6q8) ([raw file](https://github.com/robotlocomotion/drake/blob/9aabcf7da511902760120fb23e670ab3b1fb0325/systems/sensors/rgbd_renderer_vtk.cc#L132)):*
<details><summary><i>Previously, thduynguyen (Duy-Nguyen Ta) wrote…</i></summary><blockquote>

Have discussed offline with @kunimatsu. The problem is deeper than this. We need a major redesign to split this into different RGB and Depth and Label renderers for different corresponding types of cameras, which need different sets of properties. One `RenderingConfig` class is not enough to handle all these cases. For example, a depth renderer needs not only `z_near`, `z_far` but also clipping planes to handle too close/too far objects, a RGB renderer requires lighting/environment settings which other renderers don't need, and the color palette is only needed for the Label renderer.

For simplicity and minimal code changes, I'll put `RenderingConfig` and also `ColorPalette` in the base class for now. With that decision, the `Parent` pattern makes sense as `Impl` implements its parent so it needs access to all members of the parent anyway.
</blockquote></details>

BTW, thank you for leaving the note of what we discussed!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0R0X-8i00OIVwwvvuj:b-p9i4qw)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
Minor comments addressed. PTAL.
+@sherm1 for platform review please!

---

Reviewed 4 of 9 files at r3.
Review status: 7 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.h, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0QzYODXV67b7gdbt9M:-L0R5EfjqYITTloROSiJ:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/376a90dc4748fb3122290f0edac06e6576e2ab45/systems/sensors/rgbd_renderer.h#L145)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

BTW, missing a period at the end of the sentence. It would be nice to be consistent to other comments.

ditto for other comments that don't have a period.
</blockquote></details>

Fixed.

---

*[systems/sensors/rgbd_renderer.h, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0Qyg-76MSF_VnEXfMq:-L0R5JE-12_muCTmII7w:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/376a90dc4748fb3122290f0edac06e6576e2ab45/systems/sensors/rgbd_renderer.h#L162)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

BTW, TODO comment should be followed by responsible person's name(s), e.g. `TODO(thduynguyen, kunimatsu-tri) This is a ...`
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0R5SAXeY7nEQvcdXLz:bll9vuc)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending some BTWs, ptal.

---

Reviewed 1 of 10 files at r1, 8 of 9 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.cc, line 13 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RCOngtDB59z-YltJy:-L0RCOngtDB59z-YltJz:b-iumn1h) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.cc#L13)):*
> ```cpp
>     : config_(config),
>       color_palette_(kNumMaxLabel, Label::kFlatTerrain, Label::kNoBody) {
>   (void)X_WC;
> ```

BTW to avoid the "unused parameter" warning, instead use either:
- Drake's `unused(X_WC);` (in common/unused.h), or
- don't show the parameter name X_WC in the implementation function signature here (you do need it in the header so the doxygen comes out right)

---

*[systems/sensors/rgbd_renderer.h, line 73 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RB2o48LC0j1CwYT4A:-L0RB2o48LC0j1CwYT4B:bp1u407) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.h#L73)):*
> ```objc
>   ///
>   /// @param X_WC The initial pose of the renderer's unique camera viewpoint `C`
>   /// at the world coordinate system. The camera pose `C` can be updated by
> ```

BTW at -> in

---

*[systems/sensors/rgbd_renderer.h, line 158 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RBFJf--XFb8yAM2ZG:-L0RBFJf--XFb8yAM2ZH:b-1no85i) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.h#L158)):*
> ```objc
> 
>  protected:
>   /// The common configuration nedded by all implementations of this interface.
> ```

BTW nedded -> needed

---

*[systems/sensors/rgbd_renderer.h, line 165 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RBPk8-vc1fvBVmZFQ:-L0RBPk8-vc1fvBVmZFR:b-2191z3) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.h#L165)):*
> ```objc
>   /// each object/segment) hence should be moved to GeometryWorld. That would
>   /// also answer the question whether this heavy object should be a singleton.
>   ColorPalette color_palette_;
> ```

BTW our [style guide](http://drake.mit.edu/styleguide/cppguide.html#Access_Control) does not permit `protected` data members. Make these private and provide accessor methods as needed.

---

*[systems/sensors/rgbd_renderer_vtk.cc, line 124 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0REXv1xCmJasqzGW0j:-L0REXv1xCmJasqzGW0k:b-8ndz65) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer_vtk.cc#L124)):*
> ```cpp
>   void DoRenderDepthImage(ImageDepth32F* depth_image_out) const;
> 
>   void DoRenderLabelImage(ImageLabel16I* label_image_out) const;
> ```

BTW we normally use the "Do" prefix to indicate a virtual method. These are non-virtual so it is misleading to give them these names (for people who are used to our convention). Please consider removing the "Do"s here, or replacing them with "Impl" or whatever.

---

*[systems/sensors/rgbd_renderer_vtk.h, line 11 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RDciXL3aYsaezpx9F:-L0RDciXL3aYsaezpx9G:bdaj7kb) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer_vtk.h#L11)):*
> ```objc
> namespace sensors {
> 
> /// An RgbdRenderer implementation using VTK
> ```

BTW needs a period.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RFwpmB4zjE8Kv4sD2:-L0RFwpmB4zjE8Kv4sD3:b-ow2d6b) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/test/rgbd_renderer_test_util.h#L30)):*
> ```objc
>   ASSERT_NEAR(pixel[1], color.g, tolerance);
>   ASSERT_NEAR(pixel[2], color.b, tolerance);
>   ASSERT_EQ(pixel[3], alpha);
> ```

BTW normally you should use `EXPECT_NEAR` and `EXPECT_EQ` which allow testing to continue after a failure.  gtest uses `ASSERT_...` for conditions that must stop all subsequent testing, for example checking for a null pointer that must be dereferenced in subsequent tests. 

Please switch to `EXPECT` or explain why `ASSERT` should be used here (and below).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0RGxPWpQONQ-zHIw_F:b-fhb9wa)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[systems/sensors/test/rgbd_renderer_test_util.h, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RFwpmB4zjE8Kv4sD2:-L0SYeEU-zC9O6SwRATW:b-u1nscn) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/test/rgbd_renderer_test_util.h#L30)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW normally you should use `EXPECT_NEAR` and `EXPECT_EQ` which allow testing to continue after a failure.  gtest uses `ASSERT_...` for conditions that must stop all subsequent testing, for example checking for a null pointer that must be dereferenced in subsequent tests. 

Please switch to `EXPECT` or explain why `ASSERT` should be used here (and below).
</blockquote></details>

Let me answer since I'm the original author.

I'm using `ASSERT` here because this function could be called inside `for` loops to search over all the pixels in an image. See [here](https://github.com/thduynguyen/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/test/rgbd_renderer_test_util.h#L54) for the actual use case.

Would this be ok for you?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RFwpmB4zjE8Kv4sD2:-L0SxJY-5n2JJXEcPahe:boztbvz) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/test/rgbd_renderer_test_util.h#L30)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Let me answer since I'm the original author.

I'm using `ASSERT` here because this function could be called inside `for` loops to search over all the pixels in an image. See [here](https://github.com/thduynguyen/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/test/rgbd_renderer_test_util.h#L54) for the actual use case.

Would this be ok for you?
</blockquote></details>

BTW thanks, Kuni. Now that you say that I think I may have discussed it with you before -- makes sense. I think what is needed here is a comment explaining this so we don't have to keep discussing it!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RFwpmB4zjE8Kv4sD2:-L0U3YTRZgQUzn3nuAQ7:b-oor55d) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/test/rgbd_renderer_test_util.h#L30)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW thanks, Kuni. Now that you say that I think I may have discussed it with you before -- makes sense. I think what is needed here is a comment explaining this so we don't have to keep discussing it!
</blockquote></details>

I totally agree. I wonder why I haven't done that before!

@thduynguyen could you please paste my comment above to the code?  Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
Thanks @sherm1! I've fixed the code to address all your comments. PTAL.

---

Reviewed 1 of 9 files at r3, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.cc, line 13 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RCOngtDB59z-YltJy:-L0eAJtqfVLPy5_8XRbi:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.cc#L13)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW to avoid the "unused parameter" warning, instead use either:
- Drake's `unused(X_WC);` (in common/unused.h), or
- don't show the parameter name X_WC in the implementation function signature here (you do need it in the header so the doxygen comes out right)
</blockquote></details>

Fixed.

---

*[systems/sensors/rgbd_renderer.h, line 73 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RB2o48LC0j1CwYT4A:-L0e9qKYwAn3YM4TPxYz:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.h#L73)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW at -> in
</blockquote></details>

Fixed.

---

*[systems/sensors/rgbd_renderer.h, line 158 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RBFJf--XFb8yAM2ZG:-L0e9sfupMFT4r2Jrvpc:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.h#L158)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW nedded -> needed
</blockquote></details>

Fixed.

---

*[systems/sensors/rgbd_renderer.h, line 165 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RBPk8-vc1fvBVmZFQ:-L0e9unXPJiM_xTkYaC6:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer.h#L165)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW our [style guide](http://drake.mit.edu/styleguide/cppguide.html#Access_Control) does not permit `protected` data members. Make these private and provide accessor methods as needed.
</blockquote></details>

Fixed.

---

*[systems/sensors/rgbd_renderer_vtk.cc, line 124 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0REXv1xCmJasqzGW0j:-L0e9wW-Jnnn4OK23Z9U:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer_vtk.cc#L124)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW we normally use the "Do" prefix to indicate a virtual method. These are non-virtual so it is misleading to give them these names (for people who are used to our convention). Please consider removing the "Do"s here, or replacing them with "Impl" or whatever.
</blockquote></details>

Fixed.

---

*[systems/sensors/rgbd_renderer_vtk.h, line 11 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RDciXL3aYsaezpx9F:-L0eAEoVLF1NkVWH8ezW:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/rgbd_renderer_vtk.h#L11)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW needs a period.
</blockquote></details>

Fixed.

---

*[systems/sensors/test/rgbd_renderer_test_util.h, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0RFwpmB4zjE8Kv4sD2:-L0e9yf5sWzvO9Qlwqbk:bs3mhdg) ([raw file](https://github.com/robotlocomotion/drake/blob/6dbe4280dc87e7d0855d79a3e32489928b6a87b0/systems/sensors/test/rgbd_renderer_test_util.h#L30)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I totally agree. I wonder why I haven't done that before!

@thduynguyen could you please paste my comment above to the code?  Thanks!
</blockquote></details>

Comment added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0eAXZNhCtLugVI7n7k:b-y6vqm4)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
@drake-jenkins-bot mac-sierra-clang-bazel-experimental please

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0eBeH6zPJwPuw89OGe:bnuwbvt)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 1 of 9 files at r3, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Reviewed 1 of 9 files at r3, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.h, line 152 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0oWnxRnM0O9hy6Or1z:-L0oWnxS5kNNL43VU-Gr:b7rmeeq) ([raw file](https://github.com/robotlocomotion/drake/blob/b282967df84e4462fb83be75dd874943f982f81c/systems/sensors/rgbd_renderer.h#L152)):*
> ```objc
> 
>   /// Returns sky's color in RGB image.
>   const ColorI& get_sky_color() const;
> ```

You can delete this method now since the equivalent function can be called via `color_palette().get_sky_color()`.

---

*[systems/sensors/rgbd_renderer.h, line 155 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0oXAbbTBn6vm5LmxFa:-L0oXAbbTBn6vm5LmxFb:b-rqxvo7) ([raw file](https://github.com/robotlocomotion/drake/blob/b282967df84e4462fb83be75dd874943f982f81c/systems/sensors/rgbd_renderer.h#L155)):*
> ```objc
> 
>   /// Returns flat terrain's color in RGB image.
>   const ColorI& get_flat_terrain_color() const;
> ```

ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

thduynguyen commented:
Reviewed 1 of 10 files at r1, 1 of 9 files at r3, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[systems/sensors/rgbd_renderer.h, line 152 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0oWnxRnM0O9hy6Or1z:-L0ocM49uHwc41mrYVs3:b-qce3as) ([raw file](https://github.com/robotlocomotion/drake/blob/b282967df84e4462fb83be75dd874943f982f81c/systems/sensors/rgbd_renderer.h#L152)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

You can delete this method now since the equivalent function can be called via `color_palette().get_sky_color()`.
</blockquote></details>

Good point. Deleted!

---

*[systems/sensors/rgbd_renderer.h, line 155 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-L0oXAbbTBn6vm5LmxFa:-L0ocPVy_vGBo3feRGhF:b-k7mx7q) ([raw file](https://github.com/robotlocomotion/drake/blob/b282967df84e4462fb83be75dd874943f982f81c/systems/sensors/rgbd_renderer.h#L155)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

ditto.
</blockquote></details>

Deleted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7611#-:-L0q4asLkJclxtDXiseH:bg7igkv)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please.
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-memcheck-lsan please.
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-everything  please.
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-debug please.

soonho-tri commented:
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-debug please.
